### ISSUE TITLE
[Modern] Fundamental issue with relay cache/store strategy

### ISSUE THREAD

**Issue by mike-marcacci** [facebook/relay#2237]
I have a fairly common, very reasonable use case that demonstrates what I believe to be a fundamental flaw in the relay cache/store update strategy.

Let's use these simplified types in the scenario:

```graphql
type SomeObject implements Node {
    id: ID!
    name: String
    description: String
}

type SomeWrapper implements Node {
    id: ID!
    object: SomeObject
}

type Query {
  get(id: ID!): SomeWrapper
  search(term: String!): [SomeWrapper]
}
```

The important part here is that `SomeWrapper` references `SomeObject`.

In my use case, I have two `QueryRoot`s: one to fetch search results, and one to fetch all the details of a selected object. They **must be independent queries** for 2 reasons:

1. the details query fetches MUCH more data than would be practical to return on all search results
2. search query should NOT be re-run each time the selected ID changes for the details query

Here's another heavily simplified description in GraphQL (in reality the queries are large, use the `Connection` spec, and are composed of many fragments):

```graphql
query SearchQuery ($term: String!) {
    search (term: $term) {
        id
        object {
            id
            name
        }
    }
}

query DetailsQuery ($id: ID!) {
    get (id: $id) {
        id
        object {
            id
            name
            description
        }
    }
}
```

The important part here is that the `description` field is present in `DetailsQuery` but NOT in `SearchQuery`.

The problem comes when I have details pulled up for a `Wrapper` with ID `aaa`. It looks like this:

```json
{
    "id": "aaa",
    "object": {
        "id": "111",
        "name": "Name 1",
        "description": "This is a HUGE piece of data."
    }
}
```

Now, let's say that I do a search and receive results which contain `Wrapper` with ID `aaa`. It looks like this:

```json
[
    {
        "id": "aaa",
        "object": {
            "id": "111",
            "name": "Name 1"
        }
    }
]
```

Nothing has changed, and this is great. However, let's say that `aaa` has changed to reference `Object` with ID `222`:

```json
[
    {
        "id": "aaa",
        "object": {
            "id": "222",
            "name": "Name 2"
        }
    }
]
```

This becomes **EXTREMELY** problematic. Because `222` is new to relay, there is no cache for the `description` field, but the results for `DetailsQuery` are updated to reference it anyhow. While having a "description" go missing doesn't sound like a big deal, my real-world case has dozens of fields with nested queries. **Thus, receiving this kind of search result essentially makes the entire screen go blank.**

The problem, then, is that relay seems to split the middle between two more predictable updating strategies, where it would either:

1. NOT update the `DetailsQuery` results, since it knows that its cache is incomplete
2. Make a followup query to the required `node(id: ID!)` query root for the fields it knows are missing

I believe I can acheive the former by using separate environments, but this feels quite dirty.

If I recall correctly, the latter was the behavior in relay classic. Is there any way to restore this in modern?

I would be very interested to learn the reasoning behind the current behavior, and am willing to helping implement a better one if the relay team is open to the idea.

**Comment by mike-marcacci** [comment]
This is related to #1653

**Comment by mike-marcacci** [comment]
I'd like to bump this issue, as it's a pretty critical flaw and without a good understanding of *why* the relay team chose to go this direction, it's difficult for somebody in the community to address it. 

Since my issue advocates for a pretty big change, I want to make sure my ideas are compatible with your vision and goals for the project before I invest the work to make it happen. Thanks in advance!

**Comment by devknoll** [comment]
Hi Mike.

Could you provide a little more information about your scenario? When is ‘aaa.object’ being changed? Why is it changing? If it’s through a mutation, why isn’t the mutation fetching the new data? If it’s not a mutation, what’s changing it?

- Gerald

**Comment by mike-marcacci** [comment]
Hi @devknoll! Thanks so much for the followup, and sorry if the example above is a little hard to follow. The challenge here is that the node with ID `aaa` has changed on the server such that its field `object` references node `222` instead of `111`; this change isn't caused by the client who experiences the problem, but by another user or server process. Relay hasn't ever asked the server for the `description` field of `222`, but still updates the results of `DetailsQuery` with the incomplete results from the `SearchQuery` response, setting the `description` to `null`.

To be much more concise, the problem is that when there are multiple `QueryRenderer`s on a page, simply receiving the results from one can produce a cache state that is incorrect for the others. Specifically, this occurs when one query returns a `Node` that is new to the store; any referencing `Node`s are updated to point at the new `Node`, even though the necessary child fields have never been queried.

Hopefully that helps – please let me know if I can help further clarify anything!

**Comment by mike-marcacci** [comment]
Also, my question about the team's vision going forward comes from my observation that in Relay Modern, your goals included better predictability of which queries run and when they run, and to accomplish this, you moved from dynamic queries to fully-static ones. The fix for this, though, is probably going to require some kind of generated query to fill in the missing fields.

**Comment by robrichard** [comment]
@mike-marcacci can you just refetch your DetailsQuery after the second SearchQuery is executed? In your second proposal (making a follow up query), it sounds like the query you are expecting relay to generate would be almost identical to the `DetailsQuery` you already wrote.

**Comment by mike-marcacci** [comment]
Hi @robrichard, thanks so much for your comment! I think your suggestion is really great, and it's basically what I've been working on in a custom extension of `DetailsQuery`. This has the advantage of keeping everything static, while fixing the issue I describe for many use cases (particularly the simplified one I described above).

This strategy, though, doesn't work for 2 kinds of queries.

1. Queries which are very deep or contain long lists: it would be impractical to re-run the entire query every time a descendent node is changed in this way.
2. Queries for which the query's results change frequently: we don't want to disorient our users by making large spontaneously changes to result sets. (This could be addressed in userland with more deterministic modeling of queries, more durable cursors, etc.)

To me, the most "correct" thing to do is batch up `node($id)` queries with unrequested fragments/fields. Of course, I haven't actually tried to do this, and I totally understand if those who had to build/maintain this in relay classic have discovered that it just isn't worth the complexity.

I'm going to add some tests that demonstrate the issue and make some general-purpose changes to `QueryRenderer` to re-run the static query. I'm still working through distinguishing between a field has been queried and returned `null` or hasn't ever been requested, but I'll throw up a PR once it's working.

**Comment by mike-marcacci** [comment]
So I never finished writing my new `QueryRenderer` but am going to have to pick it back up. To make sure I'm conveying the severity of this issue, fields guaranteed by the API to be non-null can come become undefined because of this, making the generated flow types completely unreliable. In my case, somebody can crash an entire app just by typing into a search field, if the result list contains an updated `Node` that's used elsewhere on the screen.

I'm going to put to together a small app to demonstrate this issue, but I'm quite amazed that others don't appear to be running into this like crazy.

**Comment by mike-marcacci** [comment]
OK, I finally got around to building a full demo. Check out [this repo](https://github.com/mike-marcacci/relay-demo-2237) and [this video](https://www.youtube.com/watch?v=in84djn-oBI&feature=youtu.be).

Again, I'd like to stress the seriousness of this, as I suspect it is crashing apps across the web under very hard-to-reproduce conditions (because it is triggered by cross-user concurrency). I also want to note that I've explored a few ideas for a fix, but that there are architectural consequences to this.

**Comment by mike-marcacci** [comment]
Here are my current thoughts on how this might be fixed; I wish I could have a whiteboard session with somebody more knowledgable than me on relay's internals, but since we're just in an issue thread please take these as "first thoughts" rather than a well constructed proposal:

### Detection
This issue can be created many different ways. While my example demonstrates it when a *new* `Node` is introduced to the store, it's possible for the `Node` to already exist with insufficient data, and *later* get referenced where it doesn't belong. This makes it clear that detecting insufficient data needs to occur *when it is used*.

#### Idea 1
The most obvious way to detect missing data is to, whenever the data is provided to a `QueryRenderer`, iterate through the query (with all its fragments) and check every `Node` for the existence of each requested field.

###### Pros:
- would detect all missing fields

###### Cons:
- could get expensive on large queries with many fields
- requires that "undefined" fields are different from "null" fields in the store (might already be the case?)

#### Idea 2
Another option is to flag nodes with the query and position when its data is returned by a particular query, using a "private" client-side field. Then, whenever data is provided to a `QueryRenderer`, iterate through every returned `Node` to check the presence of the corresponding flag.

###### Pros:
- would detect all missing fields
- less expensive
- no need to differentiate between "undefined" and "null" fields in the store

###### Cons:
- a programmatically inserted Node (such as in optimistic updates) would need to specify which queries it satisfies, leaving a footgun when a new field is required
- some queries might be implicitly satisfied, resulting in unnecessary resolution

### Resolution
While detecting missing data is reasonably straightforward, resolving it isn't.

#### Idea 1
When a `QueryRender` detects missing fields, it can re-run its query to the server. This would add the missing fields.

###### Pros:
- would successfully resolve missing data
- doesn't add a ton of extra complexity

###### Cons:
- could cause unexpected "updates" if the query was time-sensitive (such as "5 most recent posts")
- could result in a lot of over-querying, if the queries are large (returning 100 `Node`s just because 1 is missing data)

#### Idea 2
When a `QueryRender` detects missing fields, it can re-request content for the specific `Node` using the `node(id: X)` API. This is similar to how relay classic worked, but the queries could be preassembled at compile time.

###### Pros:
- would successfully resolve missing data
- wouldn't cause unexpected "updates"
- wouldn't cause nearly as much querying as Idea 1

###### Cons:
- adds a serious amount of complexity

**Comment by juhaelee** [comment]
I believe I am having a similar issue not with just queryrenderer's, but also with refetch containers: https://github.com/facebook/relay/issues/2425

**Comment by sibelius** [comment]
You can use this https://github.com/facebook/relay/issues/2100#issuecomment-330034633 as a workaround

**Comment by sibelius** [comment]
Relay team is working on fragment ownership model that could probably solve this issue

> Added fragment ownership model to relay-runtime: fragments can now point to the query that owns them, which removes reliance on React Context and gives us flexibility to experiment with new apis.

**Comment by josephsavona** [comment]
> I'd like to bump this issue, as it's a pretty critical flaw and without a good understanding of _why_ the relay team chose to go this direction, it's difficult for somebody in the community to address it.

First, thanks for filing this issue. We informally call this the "missing data" scenario (an incredibly imaginative name, i know): some component fetches a field on a related object (say, a user.bestFriend.name), some update occurs where the relationship has changed (user.bestFriend is someone else now), but doesn't fetch the field, and the first component observers that user.bestFriend.name is now missing. This can definitely be surprising if you aren't expecting it and seem like a flaw: however, it's very much expected and by design.

The reason is consistency: the alternative (using the same example) is that when the update occurs and user.bestFriend has changed, the original component isn't notified and continues to show the previous friend. In other words: the UI is inconsistent. 

In client side caches there is a fundamental tradeoff between completeness (do we have all the data each component requested), consistency (do all components represent a consistent view of the world), and performance (do we achieve those properties in a reasonable amount of time). In general  a system can achieve two out of three. For example, to prioritize completeness and performance you can simply fetch each view's data in-full, up front in a single round trip, without risk that previously fetched data will go "missing" because of a consistency update. To prioritize completeness and consistency you can refetch anytime data goes missing - at the cost of possibly infinite refetches if data keeps changing on the server (and note that if you care about consistency, you can't just apply the first update bc that wouldn't be consistent - you have to wait to show the affects of an update until you know you're not missing any more data). **Relay chooses a third route: prioritize consistency and performance** - in our experience, "missing data" due to relationship changes are *relatively* more rare and also relatively easier to work around when they occur (for example, fetch the same fragment in both places to  ensure data stays in sync). 

That said, we haven't documented this tradeoff clearly anywhere so it can be surprising when it occurs. We should definitely document it somewhere - we'd appreciate feedback about where folks would intuitively expect to see this (in the architecture docs? but would enough people read those?). Also note that we're working on new react-relay APIs that use Relay's knowledge of which fields are missing (Snapshot.isMissingData is true whenever we encounter an unfetched field while reading a fragment). We're also working on better debug logging for when this case occurs.

**Comment by josephsavona** [comment]
> Relay team is working on fragment ownership model that could probably solve this issue

@sibelius You're correct that fragment ownership is related - it's part of the changes that make it easier to detect and debug this case - but it doesn't prevent this from occurring.

**Comment by jckw** [comment]
Just to confirm: this does make the auto-generated types nothing more than suggestions, right? Since really we're allowing Partial types to be returned at any point because of the "missing data" problem, which is not that much better than just using `any`.

**Comment by agg23** [comment]
This is still an ongoing discussion as Relay develops, and should not be closed.

**Comment by josephsavona** [comment]
As I [noted above](https://github.com/facebook/relay/issues/2237#issuecomment-462839992), Relay uses a normalized cache and prioritizes **consistency** and **performance** which means that data can sometimes be _incomplete_. The behavior observed here - of data for a previous query going "missing" when a subsequent query exposes relationship changes in the object graph - is an expected aspect of Relay's design that we considerable acceptable and preferable to alternatives. 

I'll update the title to reflect this, though I agree that we should keep this issue open to track improvements to the docs and APIs that allow developers to mitigate this in the cases when it is problematic.

**Comment by trey-miller** [comment]
Hi I know this issue is a little old by now, but it's introducing insidious bugs in our code, where we rely on the types being accurate. It seems odd to me to delete defined field data if some other query happens to load the same object but didn't request that field. Why not keep the previous field's data if not requested by the subsequent query?

Relay knows which fields are requested, so if a field wasn't requested, why assume it's undefined instead of just, well, not requested, and then keep using the old data on the field if that field was not requested? Isn't it a bit presumptuous to assume the data has changed on the server to delete data in a field, simply because a later query didn't request that specific field?

On the one hand, you could end up showing stale data, but on the other (current) hand, you get subtle bugs causing catastrophic runtime errors...
